File: app/src/main/java/com/example/foodapp/ui/fragments/SearchFragment.java
package com.example.foodapp.ui.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.PopupMenu;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.SearchView;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.view.MenuCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import com.example.foodapp.R;
import com.example.foodapp.viewmodel.RecipeViewModel;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class SearchFragment extends Fragment {

    private SearchView searchSv;
    private ImageView userIconIv;
    private ConstraintLayout layout;
    private RecipeViewModel recipeViewModel;
    private List<OnUserMenuItemClickListener> onUserMenuItemCLickListenerList = new ArrayList<>();

    public void setOnUserMenuItemCLickListenerList(@NonNull OnUserMenuItemClickListener onUserMenuItemCLickListenerList) {
        this.onUserMenuItemCLickListenerList.add(onUserMenuItemCLickListenerList);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        layout = (ConstraintLayout) inflater.inflate(R.layout.fragment_search, container, false);
        recipeViewModel = new ViewModelProvider(requireActivity()).get(RecipeViewModel.class);
        bindView();
        bindAction();
        return layout;
    }

    private Consumer<String> onSearchListener;

    public void setOnSearchListener(Consumer<String> onSearchListener) {
        this.onSearchListener = onSearchListener;
    }

    private void bindView() {
        searchSv = layout.findViewById(R.id.searchSv);
        userIconIv = layout.findViewById(R.id.userIconIv);
    }

    private void bindAction() {
        searchSv.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                recipeViewModel.setSearchParams(query, null, null, null, null);
                if (onSearchListener != null) onSearchListener.accept(query);
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                return false;
            }
        });
        userIconIv.setOnClickListener(view -> {
                    PopupMenu popup = new PopupMenu(getContext(), view);
                    popup.getMenuInflater().inflate(R.menu.user_menu, popup.getMenu());
                    MenuCompat.setGroupDividerEnabled(popup.getMenu(), true);
                    popup.show();
                    popup.setOnMenuItemClickListener(menuItem -> {
                        switch (menuItem.getItemId()) {
                            case R.id.homeI:
                                onUserMenuItemCLickListenerList.forEach(OnUserMenuItemClickListener::onHomeClick);
                                return true;
                            case R.id.savedRecipesI:
                                onUserMenuItemCLickListenerList.forEach(OnUserMenuItemClickListener::onSavedRecipesClick);
                                return true;
                            case R.id.userSettingsI:
                                onUserMenuItemCLickListenerList.forEach(OnUserMenuItemClickListener::onUserSettingsClick);
                                return true;
                            case R.id.logOutI:
                                onUserMenuItemCLickListenerList.forEach(OnUserMenuItemClickListener::onLogOutClick);
                                return true;
                            default:
                                return false;
                        }
                    });
                }
        );
    }

    public static class OnUserMenuItemClickListener {
        public void onHomeClick() {
        }

        public void onSavedRecipesClick() {
        }

        public void onUserSettingsClick() {
        }

        public void onLogOutClick() {
        }
    }
}

File: app/src/main/java/com/example/foodapp/ui/fragments/SearchFragment.kt
package com.example.foodapp.ui.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.PopupMenu
import androidx.appcompat.widget.SearchView
import androidx.core.view.MenuCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import com.example.foodapp.R
import com.example.foodapp.databinding.FragmentSearchBinding
import com.example.foodapp.viewmodel.RecipeViewModel
import dagger.hilt.android.AndroidEntryPoint
import java.util.function.Consumer

@AndroidEntryPoint
class SearchFragment : Fragment() {

    private lateinit var binding: FragmentSearchBinding
    private val recipeViewModel: RecipeViewModel by viewModels({ requireActivity() })
    var onUserMenuItemCLickListener: OnUserMenuItemClickListener? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentSearchBinding.inflate(inflater, container, false)
        bindAction()
        return binding.root
    }

    private var onSearchListener: Consumer<String>? = null
    fun setOnSearchListener(onSearchListener: Consumer<String>?) {
        this.onSearchListener = onSearchListener
    }

    private fun bindAction() {
        binding.searchSv.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                recipeViewModel.setSearchParams(query, null, null, null, null)
                if (onSearchListener != null) onSearchListener!!.accept(query)
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                return false
            }
        })
        binding.userIconIv.setOnClickListener { view: View? ->
            val popup = PopupMenu(context, view)
            popup.menuInflater.inflate(R.menu.user_menu, popup.menu)
            MenuCompat.setGroupDividerEnabled(popup.menu, true)
            popup.show()
            popup.setOnMenuItemClickListener { menuItem: MenuItem ->
                when (menuItem.itemId) {
                    R.id.homeI -> {
                        onUserMenuItemCLickListener?.onHomeClick()
                        return@setOnMenuItemClickListener true
                    }

                    R.id.savedRecipesI -> {
                        onUserMenuItemCLickListener?.onSavedRecipesClick()
                        return@setOnMenuItemClickListener true
                    }

                    R.id.userSettingsI -> {
                        onUserMenuItemCLickListener?.onUserSettingsClick()
                        return@setOnMenuItemClickListener true
                    }

                    R.id.logOutI -> {
                        onUserMenuItemCLickListener?.onLogOutClick()
                        return@setOnMenuItemClickListener true
                    }

                    else -> return@setOnMenuItemClickListener false
                }
            }
        }
    }

    open class OnUserMenuItemClickListener {
        open fun onHomeClick() {}
        open fun onSavedRecipesClick() {}
        open fun onUserSettingsClick() {}
        open fun onLogOutClick() {}
    }
}